### PR TITLE
[#159133590] Code refactoring

### PR DESCRIPTION
**Pull request by andela-Taiwo**
#### What does this PR do?
- Code refactoring
#### Description of task completed
- add resource controllers for user and admin
- update tests request parameters to use factories
- fix bug the flash messages
#### How should it be tested?
- `rspec`
#### Related PivotalTracker story
 - #159133590

### CODE REVIEW

File: app/app.rb [hunk 1]
```diff
@@ -110,14 +112,13 @@ def html(view)
 		"
   end
 
-  get '/test/add_job' do
+  get "/test/add_job" do
     check_admin_authentication
     "
-		<p>Added Job: #{ResourceNotification.set(:queue => :default).perform_async}</p>
+		<p>Added Job: #{ResourceNotification.set(queue: :default).perform_async}</p>
```
Comment (morriskimani) [inline app/app.rb]: Let's have this in a view

File: app/controllers/admin_topic_controller.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+require "sinatra/flash"
+require "pony"
+
+module Sinatra
+  module App
+    module TopicController
+      def self.registered(app)
+
+        app.before do
+          check_admin_authentication if request.path_info == "/admin/"
```
Comment (morriskimani) [inline app/controllers/admin_topic_controller.rb]: :+:

File: app/controllers/admin_topic_controller.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        end
+        app.post "/admin/topic" do
+          @topic = Topic.new(params[:topic])
+          if @topic.save
+            flash[:success] = "Successfully add a new topic"
+            redirect "/admin/topic"
```
Comment (morriskimani) [inline app/controllers/admin_topic_controller.rb]: Let's have this outside the if block

File: app/controllers/admin_topic_controller.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+          @title = "Topic"
+          erb :'topics/topic_form'
+        end
+
+        app.get "/admin/topic/:id" do
+          @topic = Topic.find_by_id(params[:id])
+          erb :'topics/edit_topic_form' if @topic
+        end
+
+        app.patch "/admin/topic/:id/edit" do
```
Comment (morriskimani) [inline app/controllers/admin_topic_controller.rb]: Do we need `edit` in the route name? The HTTP method used already gives an indication of the action that takes place

File: app/controllers/admin_topic_controller.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+          @topic = Topic.find_by_id(params[:id])
+          if @topic
+            erb :'topics/edit_topic_form'
```
Comment (morriskimani) [inline app/controllers/admin_topic_controller.rb]: Why do we need to render this form here?

File: app/controllers/admin_topic_controller.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            else
+              error = @topic.errors.messages
+              flash[:error] = error
+            end
+            redirect "/admin/topic"
+          end
+        end
+
+        app.delete "/admin/topic/:id/delete" do
+          @topic = Topic.find_by_id(params[:id])
```
Comment (morriskimani) [inline app/controllers/admin_topic_controller.rb]: Let's DRY out this controller by using a before filter to set this instance variable

File: app/controllers/resource_controller.rb [hunk 7]
```diff
@@ -1,83 +1,71 @@
...
+        app.patch "/admin/resource/:id/edit" do
           @resource = Resource.find_by_id(params[:id])
           if @resource
             erb :'resources/edit_resource_form'
-
-            @resource.title = params[:title] if params.has_key?('title')
-            @resource.description = params[:description] if params.has_key?('description')
-            @resource.url = params[:url] if params.has_key?('url')
-
+            @resource.title = params[:title].strip if params.has_key?("title")
```
Comment (emugaya) [inline app/controllers/resource_controller.rb]: Consider using `resource.update_attributes(params)`

File: spec/controllers/session_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+
+RSpec.describe Sinatra::App:: LoginSession do
+  include Rack::Test::Methods
+  include Warden::Test::Helpers
+  let(:user) { create :user }
+  let(:admin)  { build :admin }
+  after { Warden.test_reset! }
+  def app
```
Comment (morriskimani) [inline spec/controllers/session_spec.rb]: I think it might be better to have this as a  helper method

**Review by morriskimani — COMMENTED** [review]
Good work!
Just some few comments

**Pull request merged**